language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -129,452 +110,192 @@ impl AbstractState {
-                                .union(&nonce_set)
-                                .cloned()
-                                .collect::<BTreeSet<_>>();
-                            if !borrowed_nonce_set_update.is_empty() {
-                                new_index_to_nonce_set
-                                    .insert(idx.clone(), borrowed_nonce_set_update);
-                            }
-                        } else {
-                            new_index_to_nonce_set.insert(idx.clone(), borrowed_nonce_set.clone());
-                        }
-                    }
-                    if !new_index_to_nonce_set.is_empty() {
-                        new_borrows.insert(
-                            *src_nonce,
-                            BorrowInfo::FieldsBorrowedBy(new_index_to_nonce_set),
-                        );
-                    }
-                }
-            }
-        }
+    /// gets all borrows of a global
+    pub fn global_borrows(&self, idx: StructDefinitionIndex) -> BTreeSet<Nonce> {
+        self.borrow_graph
+            .consistent_borrows(self.frame_root, LabelElem::Global(idx))
+    }
 
-        self.locals = new_locals;
-        self.globals = new_globals;
-        self.borrows = new_borrows;
-        self.partition.remove_nonce(nonce);
+    /// gets all borrows of `nonce` that might be consistent with a field
+    pub fn field_borrows(&self, nonce: Nonce, idx: FieldDefinitionIndex) -> BTreeSet<Nonce> {
+        self.borrow_graph
+            .consistent_borrows(nonce, LabelElem::Field(idx))
     }
 
-    /// checks if there are any pending borrows on value
-    pub fn is_full(&self, value: &AbstractValue) -> bool {
-        match value {
-            AbstractValue::Reference(nonce) => !self.borrows.contains_key(&nonce),
-            AbstractValue::Value(_, nonce_set) => nonce_set.is_empty(),
-        }
+    /// checks if `nonce` is borrowed
+    pub fn is_nonce_borrowed(&self, nonce: Nonce) -> bool {
+        !self.borrow_graph.all_borrows(nonce).is_empty()
     }
 
-    /// returns the set of nonces borrowing from nonce that might alias some idx-extension of nonce
-    pub fn borrowed_nonces_for_field(
-        &self,
-        idx: FieldDefinitionIndex,
-        nonce: Nonce,
-    ) -> BTreeSet<Nonce> {
-        if self.borrows.contains_key(&nonce) {
-            match &self.borrows[&nonce] {
-                BorrowInfo::BorrowedBy(borrowed_by) => borrowed_by.clone(),
-                BorrowInfo::FieldsBorrowedBy(fields_borrowed_by) => {
-                    if fields_borrowed_by.contains_key(&idx) {
-                        fields_borrowed_by[&idx].clone()
-                    } else {
-                        BTreeSet::new()
-                    }
-                }
-            }
-        } else {
-            BTreeSet::new()
-        }
+    /// checks if `nonce` is borrowed via an edge with the empty label
+    pub fn is_nonce_nil_borrowed(&self, nonce: Nonce) -> bool {
```

Comment:
This abstraction feels a bit leaky to me.  I feel that the abstract state should really only expose APIs like `is_readable`, `is_writable`, `is_freezable` etc. And none of the internal notions of the borrow graph should be exposed to the transfer function logic

Reply:
I have been wrestling with this issue.  Currently, AbstractState is not aware of the mutable/immutable tag of a reference.  I am now looking into how to cleanly make this information available to AbstractState.  I agree that the APIs you are suggesting are cleaner.

Reply:
I have struggled with it too, while there is a bit too much delegation, I found I reached a good abstraction over the core APIs in my formulation: 

https://github.com/libra/libra/blob/master/language/move-lang/src/cfgir/borrows/state.rs#L171#L233

Reply:
My refactoring has sorted the issue of leakiness.  I have moved the nonce counter from TypeAndMemorySafetyAnalysis to AbstractState.  This worked wonders in hiding details of the borrow graph from the transfer function propagation.